File: app/src/test/java/com/example/myapplication/core/program/ScheduleReschedulerTest.kt
package com.example.myapplication.core.program

import java.time.DayOfWeek
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

class ScheduleReschedulerTest {
    private val sessions = listOf(
        ReschedulableSession(1, 0, 100, completedEpochDay = 100, demanding = true),
        ReschedulableSession(2, 1, 102, completedEpochDay = null, demanding = true),
        ReschedulableSession(3, 2, 104, completedEpochDay = null, demanding = true),
        ReschedulableSession(4, 3, 107, completedEpochDay = null, demanding = false),
    )

    @Test
    fun `moves pending session earlier and never changes completed rows`() {
        val preview = ScheduleRescheduler.preview(
            sessions, 2, 101, todayEpochDay = 100,
            trainingDays = setOf(day(101), day(103), day(106)),
        )

        assertEquals(listOf(2L, 3L, 4L), preview.changes.map { it.sessionId })
        assertTrue(preview.changes.none { it.sessionId == 1L })
        assertEquals(listOf(101L, 103L, 106L), preview.changes.map { it.newEpochDay })
    }

    @Test
    fun `moves later over week boundary while preserving sequence`() {
        val preview = ScheduleRescheduler.preview(
            sessions, 2, 108, todayEpochDay = 100,
            trainingDays = setOf(day(108), day(110), day(113)),
        )

        assertEquals(listOf(108L, 110L, 113L), preview.changes.map { it.newEpochDay })
        assertEquals(preview.changes.sortedBy { it.newEpochDay }, preview.changes)
    }

    @Test
    fun `warns when anchor is not a selected training weekday`() {
        val preview = ScheduleRescheduler.preview(
            sessions, 2, 103, todayEpochDay = 100,
            trainingDays = setOf(day(102), day(104)),
        )

        assertTrue(preview.warningsVi.any { "ngày tập" in it })
        assertEquals(103L, preview.changes.first().newEpochDay)
    }

    @Test
    fun `warns for consecutive demanding sessions`() {
        val preview = ScheduleRescheduler.preview(
            sessions, 2, 101, todayEpochDay = 100,
            trainingDays = setOf(day(101), day(102), day(105)),
        )

        assertTrue(preview.warningsVi.any { "liên tiếp" in it })
    }

    @Test
    fun `rejects past date and arithmetic overflow`() {
        assertThrows(IllegalArgumentException::class.java) {
            ScheduleRescheduler.preview(sessions, 2, 99, 100, setOf(day(100)))
        }
        assertThrows(ArithmeticException::class.java) {
            ScheduleRescheduler.preview(
                sessions,
                2,
                Long.MAX_VALUE,
                Long.MAX_VALUE,
                setOf(DayOfWeek.MONDAY),
            )
        }
    }

    private fun day(epochDay: Long): DayOfWeek = java.time.LocalDate.ofEpochDay(epochDay).dayOfWeek
}

File: flutter/test/core/program/schedule_rescheduler_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_app/core/model/goal_models.dart';
import 'package:gym_app/core/program/schedule_rescheduler.dart';

void main() {
  group('ScheduleReschedulerTest', () {
    final sessions = [
      ReschedulableSession(sessionId: 1, sequenceIndex: 0, dueEpochDay: 100, completedEpochDay: 100, demanding: true),
      ReschedulableSession(sessionId: 2, sequenceIndex: 1, dueEpochDay: 102, completedEpochDay: null, demanding: true),
      ReschedulableSession(sessionId: 3, sequenceIndex: 2, dueEpochDay: 104, completedEpochDay: null, demanding: true),
      ReschedulableSession(sessionId: 4, sequenceIndex: 3, dueEpochDay: 107, completedEpochDay: null, demanding: false),
    ];

    WeekDay day(int epochDay) {
      final date = DateTime.fromMillisecondsSinceEpoch(epochDay * 24 * 60 * 60 * 1000, isUtc: true);
      return WeekDay.fromValue(date.weekday);
    }

    test('moves pending session earlier and never changes completed rows', () {
      final preview = ScheduleRescheduler.preview(
        sessions: sessions,
        selectedSessionId: 2,
        newEpochDay: 101,
        todayEpochDay: 100,
        trainingDays: {day(101), day(103), day(106)},
      );

      expect(preview.changes.map((c) => c.sessionId).toList(), equals([2, 3, 4]));
      expect(preview.changes.any((c) => c.sessionId == 1), isFalse);
      expect(preview.changes.map((c) => c.newEpochDay).toList(), equals([101, 103, 106]));
    });

    test('moves later over week boundary while preserving sequence', () {
      final preview = ScheduleRescheduler.preview(
        sessions: sessions,
        selectedSessionId: 2,
        newEpochDay: 108,
        todayEpochDay: 100,
        trainingDays: {day(108), day(110), day(113)},
      );

      expect(preview.changes.map((c) => c.newEpochDay).toList(), equals([108, 110, 113]));
      final sortedChanges = List<SessionDateChange>.from(preview.changes)
        ..sort((a, b) => a.newEpochDay.compareTo(b.newEpochDay));
      expect(preview.changes, equals(sortedChanges));
    });

    test('warns when anchor is not a selected training weekday', () {
      final preview = ScheduleRescheduler.preview(
        sessions: sessions,
        selectedSessionId: 2,
        newEpochDay: 103,
        todayEpochDay: 100,
        trainingDays: {day(102), day(104)},
      );

      expect(preview.warningsVi.any((w) => w.contains("ngày tập")), isTrue);
      expect(preview.changes.first.newEpochDay, equals(103));
    });

    test('warns for consecutive demanding sessions', () {
      final preview = ScheduleRescheduler.preview(
        sessions: sessions,
        selectedSessionId: 2,
        newEpochDay: 101,
        todayEpochDay: 100,
        trainingDays: {day(101), day(102), day(105)},
      );

      expect(preview.warningsVi.any((w) => w.contains("liên tiếp")), isTrue);
    });

    test('rejects past date and arithmetic overflow', () {
      expect(
        () => ScheduleRescheduler.preview(
          sessions: sessions,
          selectedSessionId: 2,
          newEpochDay: 99,
          todayEpochDay: 100,
          trainingDays: {day(100)},
        ),
        throwsArgumentError,
      );

      const int maxInt = 9223372036854775807;
      expect(
        () => ScheduleRescheduler.preview(
          sessions: sessions,
          selectedSessionId: 2,
          newEpochDay: maxInt,
          todayEpochDay: maxInt,
          trainingDays: {WeekDay.monday},
        ),
        throwsA(isA<UnsupportedError>()),
      );
    });
  });
}
